### PR TITLE
Save pages returned by `oa_request()` and do not process further

### PR DESCRIPTION
**Pull request by rkrug**
This is inspired by #166 

It is. not possible, due to memory limitations, to download a large number of works (thousands, millions). This pull request introduces an argument `save_pages` to the function `oa_request` which, if provided, saves all downloaded pages in the directory specified by the argument `save_pages`. No further processing of the downloaded pages is done, as this would again lead to memory problems, a vector containing the file names of the pages is returned.

This is at the moment a very rudimentary implementation, and could be tweaked further for performance, but I think this is not needed as a quick profiling indicated that the API calls take about 80% of the time, while the `saveRDS()` only 10%.

### CODE REVIEW

File: R/oa_fetch.R [hunk 1]
```diff
@@ -163,6 +163,15 @@ oa_fetch <- function(entity = if (is.null(identifier)) NULL else id_type(shorten
 #' Either "cursor" for cursor paging or "page" for basic paging.
 #' When used with options$sample, please set `paging = "page"`
 #' to avoid duplicates.
+#' @param save_pages Character.
```
Comment (yjunechoe) [inline R/oa_fetch.R:166]: Two quick thoughts about `save_pages`:

1) I think the name could be a little clearer - people might confuse this with a toggle option taking T/F. Maybe something like `out_dir` that makes it explicit that it expects a directory path as value?

2) The argument description should ideally be concise and document just its function. The motivation and other notes (about memory issues, etc.) could move to its own `@section` or `@notes`

File: R/oa_fetch.R [hunk 2]
```diff
@@ -361,6 +373,17 @@ oa_request <- function(query_url,
   # Setting items per page
   query_ls[["per-page"]] <- per_page
 
+  # Setup save_pages if not NULL
+  if (!is.null(save_pages)) {
+    if (!dir.exists(save_pages)) {
+      dir.create(save_pages)
+    }
+    save_pages <- normalizePath(save_pages)
```
Comment (yjunechoe) [inline R/oa_fetch.R:381]: Do we perhaps want to `normalizePath()` before we potentially `dir.create()`?

**Review by yjunechoe — COMMENTED** [review]
Thanks for the PR! I left a few minor (design-related) comments for your consideration (haven't tested out how the code runs yet).

File: R/oa_fetch.R (hunk 2 above)
Comment (rkrug) [reply R/oa_fetch.R:381]: Thought that as well - but it creates a warning if the directory dow=es not exists. But one could use `mustWork = FALSE`?

Comment (yjunechoe) [reply R/oa_fetch.R:381]: I've never used that argument before but I think that makes sense - basically, I think we'd prefer the path used for `dir.create()` and stored in `save_pages` to be exactly identical.

File: R/oa_fetch.R (hunk 1 above)
Comment (rkrug) [reply R/oa_fetch.R:166]: 1. Agreed. `out_dir` sounds a bit like where to save the output - which it is not. Rather `output_pages_to`?
2. OK - I will streamline it.

Comment (yjunechoe) [reply R/oa_fetch.R:166]: Yeah `output_pages_to` sounds great!

Comment (rkrug) [reply R/oa_fetch.R]: Changed in last commit

**Review by trangdata — CHANGES_REQUESTED** [review]
I'm not sure it's optimal to use `saveRDS` within the package. 

I'm currently working on a different branch on improving paging in `oa_fetch` and `oa_request`. Let me finish some final touches and perhaps you can help me test my implementation @rkrug? The idea is to use basic paging for fetching individual pages. Personally, I think this change would be a little less disruptive than `saveRDS`. I will give some examples in the PR. And whether we move forward with `output_pages_to` or not, I think my branch would still improve the paging implementation. 

Would love to hear your thoughts @rkrug @yjunechoe. Stay tuned!

**Comment by rkrug** [comment]
From my experience with snowballing starting from more than 3000 papers, the limitation is really memory, i.e. the memory management of R. 
If one would download one page, then reserve the memory needed for one object to hold all data from all pages (e.g. `rep(ONE_PAGE_DATA_structure, n_pages`) by creating a dummy object and replacing after each page the page in the object, it could work - unless the final object is to large (and this would speed the process up as only one memory allocation is needed). 
But still - I guess that the total memory limit kicks in sooner or later. So the fallback solution of saving each page using `saveRDS` would be a safety net, which one could use if a) an argument is set, or b) the object containing the data I can not be created.


Just my thoughts, and looking forward to your pull request, @trangdata

File: R/oa_fetch.R [hunk 3]
```diff
@@ -312,6 +322,7 @@ oa_request <- function(query_url,
                        per_page = 200,
                        paging = "cursor",
                        pages = NULL,
+                       output_pages_to = NULL,
                        count_only = FALSE,
```
Comment (rkrug) [inline R/oa_fetch.R:326]: This is not more than a proof of concept implementation.

**Comment by trangdata** [comment]
> I'm also a little wary of using saveRDS. It's a weird middle ground between where you apply some but not all processing to the data received from OA.

@yjunechoe Not sure you saw [my comment](https://github.com/ropensci/openalexR/pull/183#issuecomment-1775085348) in my Improve paging control PR, but I'll reiterate it here:  at the moment, I think it's out of scope for openalexR to handle any type of serialization/IO. We currently provide enough flexibility for the user to do this themselves. 

In general, I think making many requests to the Open Alex database is not a common use case, and I would generally advise against it. One can [download the snapshot](https://docs.openalex.org/download-all-data/download-to-your-machine) themselves and proceed with the filtering/processing without needing openalexR.

Therefore, my current suggestion would be to close this PR.

**Comment by yjunechoe** [comment]
Ah I'd missed the discussion going over there! Yes, I don't disagree with just closing this as out of scope (my comment was more of a "if we do implement this" comment)

**Comment by yjunechoe** [comment]
Closing this as out of scope for the package, in favor of #183 and the continuing discussion on #182 (where we can resolve how to streamline this workflow without introducing too much maintenance burden). Apologies for my earlier confusion!

**Pull request closed without merging**